### ISSUE TITLE
[arch] grounding caso<->dataset: case_architect debe declarar dataset_schema_required y data_generator debe validarlo

### ISSUE THREAD

**Issue by JuanCamiloDorado** [4Tech-Labs/ADAM-EDU#225]
# [arch] grounding caso <-> dataset: las columnas mencionadas en el dilema deben existir en el dataset generado

## Contexto

Detectado durante dogfood real del caso "LogiTech Solutions — Crisis de retencion y el salto a la logistica predictiva" (caso ml_ds, harvard_with_eda).

## Sintoma

El caso M1 plantea un dilema operativo concreto:

> "Predecir el riesgo de incumplimiento en plazos de entrega para renovaciones de contratos minoristas de alta criticidad."

Y solicita en M3 algoritmos como `XGBoost con tuning` y `Logistic Regression + Random Forest + SHAP` para clasificar "riesgo de retraso".

Sin embargo, el dataset generado por `data_generator` solo contenia columnas financieras / CRM:

```
nps, costs, period, revenue, categoria, churn_rate, margin_pct,
customer_ltv, retention_m1/m3/m6/m12, engagement_score, transaction_amount
```

**No habia ninguna columna operativa de entregas, retrasos, plazos, tiempos de transito ni nada relacionado con la prediccion solicitada.**

Resultado:
- El Architect Engineer (M3) alucino la variable "tiempos" porque la necesitaba para defender la hipotesis.
- El notebook M3, al no encontrar la columna, hizo fallback heuristico a `categoria` (una categorica generica de 5 clases) como proxy del target — entrenando un modelo que no responde la pregunta del caso.
- Pedagogicamente roto: el alumno entrenara algoritmos sobre el dataset equivocado y obtendra metricas que no significan nada respecto al dilema.

## Causa raiz

Hoy en el grafo no existe contrato explicito caso <-> dataset:
- `case_architect` produce el dilema y los algoritmos sugeridos sin declarar que columnas operativas son indispensables.
- `data_generator` produce un dataset siguiendo un esquema generico (KPIs de negocio) sin verificar que exista al menos una columna alineada con el dilema.
- `m3_notebook_generator` recibe ambos sin manera de detectar el mismatch y cae en fallback silencioso.

## Propuesta (alto nivel — requiere ADR)

1. **Contrato `dataset_schema_required` en `case_architect`**: cuando produce el caso, declara una lista minima de columnas semanticamente requeridas por el dilema (con tipo y rol — feature operativa / target candidato / dimension temporal).
2. **Validador en `data_generator`**: antes de entregar el dataset, verifica que el schema generado cubre al menos el `dataset_schema_required.target_role` y al menos una `feature_role` operativa. Si no, regenera con prompt aumentado o degrada explicitamente el caso.
3. **Senal de mismatch en M2**: si la cobertura es parcial, el reporte EDA M2 lo declara como `data_gap` y M3 lo arrastra como nota metodologica explicita en lugar de inventar variables.

## Test cases esperados

- Caso "logistica predictiva" -> dataset incluye columna de fecha de entrega + columna de status/retraso + features operativas (volumen, region, etc.).
- Caso "churn prediction" -> dataset incluye al menos una flag de churn binaria + features de uso.
- Caso "fraud detection" -> dataset incluye amount + timestamp + flag de fraude (con desbalance realista).

## Por que no se arregla en el PR de hardening del notebook

PR #224 (M3 notebook quality hardening) parchea sintomas downstream (forzar metricas, evitar leakage, prohibir alucinacion de columnas en hipotesis), pero el problema raiz esta upstream. Arreglarlo aqui requeriria:
- Cambiar el contrato de salida de `case_architect`.
- Agregar validacion + posible re-prompting en `data_generator`.
- Revisar el flow de degradacion cuando el dataset no se puede alinear.

Eso es ADR-level y no cabe en un cambio quirurgico pre-demo.

## Severidad

Alta. Mientras el mismatch persista, cualquier caso ml_ds donde el dilema mencione variables operativas / dominio especifico va a producir notebooks que entrenan sobre el dataset equivocado.

## Referencias

- PR #224 (parche downstream).
- Caso de prueba: docs/dataset_LogiTech_Solutions_*.csv + notebook generado.
- AGENTS.md `Sensitive Areas`: `backend/src/case_generator/**`.

**Comment by JuanCamiloDorado** [comment]
## Refuerzo con segundo dogfood — caso "LogiSol — El dilema predictivo en la última milla" (28 abr 2026)

Re-ejecuté el flujo completo después de mergear PR #224 + PR #226 (4 fixes quirúrgicos al prompt M3). **Los fixes downstream funcionan tal como prometido**, pero confirman que el problema raíz sigue siendo este issue (#225) — el grounding caso↔dataset.

### Lo que SÍ mejoró (verificado en el notebook generado)

- Sección 2.1 ahora detecta `period` correctamente (date_aliases ampliado).
- Time-split temporal cronológico real activado: `Split temporal por period → train hasta 2024-XX`.
- Imputación split-first sin leakage estadístico: `med = X_train.median(); X_train.fillna(med); X_test.fillna(med)` en ambos bloques (RF y XGB).
- XGBoost ya no rompe: `XGBClassifier(random_state=42, eval_metric='logloss')` sin `use_label_encoder`, con `except (ImportError, TypeError, AttributeError)` y fallback limpio a `GradientBoostingClassifier`.

### Lo que sigue roto upstream (este issue) — nueva evidencia

**Caso LogiSol** plantea: *"riesgo de incumplimiento en plazos de entrega de última milla"*.
**Dataset generado por `data_generator`** (n=200, 14 columnas):

```
nps, costs, label (cat_1..cat_5), period, revenue, churn_rate, margin_pct,
customer_ltv, retention_m1, retention_m3, retention_m6, retention_m12,
engagement_score, transaction_amount
```

Mismo patrón que LogiTech: **cero columnas operativas de logística / entregas / retrasos / SLAs**. El caso pide predecir incumplimiento de entregas y el dataset es 100% KPIs SaaS/CRM. El notebook cae en fallback heurístico a `label` (cat_1..cat_5, ruido sintético sin causalidad con las features).

### Modos de fallo NUEVOS detectados en este dogfood (a considerar al diseñar el ADR)

1. **Leakage temporal por construcción de features**: el dataset incluye `retention_m1, retention_m3, retention_m6, retention_m12` como columnas. Cuando el notebook hace time-split por `period`, esas columnas son por definición señales del futuro respecto a su propio timestamp (la retención a 12 meses del periodo `2023-01` es información medida en `2024-01`). Incluirlas como features en el train rompe la validez del split aun cuando el split mismo sea cronológicamente correcto. **Implicación para el contrato**: `dataset_schema_required` debe declarar no solo qué columnas existen sino el `temporal_offset_months` de cada feature respecto a la observación, y `data_generator` debe excluir o etiquetar como leak-prone las features con offset > 0.

2. **Target sintético sin relación causal con features**: `label = cat_1..cat_5` aparenta ser un target categórico válido, pero al inspeccionar el CSV no hay patrón discriminable contra ninguna feature. El notebook entrenará y reportará F1 macro ≈ 0.20 (random sobre 5 clases balanceadas). **Implicación**: `data_generator` debería inyectar señal mínima detectable entre target y al menos una feature (relación lineal o no lineal débil pero presente) o, si el target es genuinamente ruido, declararlo en metadata para que M2 lo reporte como `data_quality_warning`.

3. **Hipótesis pedagógica desalineada del experimento**: el markdown del bloque RF dice *"Si el cliente presenta una caída temprana en el engagement_score, entonces la churn_rate debería cambiar..."* pero el modelo entrena contra `label`, no `churn_rate`. Síntoma del mismo desacople: `case_architect` redacta hipótesis sobre variables que no son el target real que el notebook va a usar. **Implicación**: el `dataset_schema_required` debe declarar `target_column` explícitamente y el M3 prompt debe forzar al notebook a usar ESE target o emitir `REQUISITO FALTANTE` explícito (sin fallback silencioso a "última columna categórica").

### Modos de fallo adyacentes (prompt-level, NO bloquean este issue, los anoto para no perderlos)

- `df.sample(min(len(df), 5000), random_state=42)` ANTES de `sort_values(date_col)` es safe para n≤5000 pero rompe el orden cronológico para datasets más grandes. Patrón a corregir post-demo en `M3_NOTEBOOK_ALGO_PROMPT`.
- Regla K (size guard) solo dispara con `n < 50` total. No contempla `test_size_per_class < 10` (con n=200, 5 clases, 80/20 split → ~8 ejemplos por clase en test, F1 macro con varianza inestimable). Mejora futura del prompt.

### Recomendación de scope para el ADR

Considerar al menos estos 4 contratos en el ADR del grounding:

| Campo en `dataset_schema_required` | Propósito |
|---|---|
| `target_column` (name, type, role) | Forzar alineación caso ↔ notebook ↔ hipótesis |
| `feature_columns[]` (name, type, `temporal_offset_months`) | Detectar leakage por construcción antes de servir el dataset |
| `domain_features_required[]` (semantic_tag) | "logistics_timestamp", "delivery_status", etc. — `data_generator` debe materializar al menos una |
| `min_signal_strength` (target ↔ features) | Validar post-generación que el target no es ruido puro |

### Referencias

- PR #224 (M3 hardening base).
- PR #226 (4 fixes quirúrgicos: date_aliases, split-first imputation, deprecated APIs blacklist, n<50 guard).
- Caso LogiSol: `docs/dataset_LogiSol_—_El_dilema_predictivo_en_la_última_milla.csv` + `docs/notebook_LogiSol_El_dilema_predictivo_en_la_última_milla.ipynb`.
- Caso LogiTech (dogfood original): referenciado en el cuerpo del issue.

**Severidad confirmada como Alta**: dos casos independientes (LogiTech, LogiSol) reproducen el mismo mismatch. No es un edge case del primer caso de prueba.
